### PR TITLE
Add STM32H5 flash driver

### PR DESCRIPTION
**Pull request by Luctins**
This was based on the stm32h7 driver minus the flash bank logic and the fact that the h5 has secure and non secure flash.

The current implementation works, but may have some outstanding issues, like the secure registers and register lockouts, so it needs a bit more testing to be considered "ready". 

I've implemented this so I could do DFU on the stm32h5, and for that it works.

Any review is very much appreciated.

### CODE REVIEW

File: embassy-stm32/src/flash/h5.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+    cortex_m::asm::isb();
+    cortex_m::asm::dsb();
+    fence(Ordering::SeqCst);
+
+    clear_all_err();
+
+    pac::FLASH.nscr().write(|w| {
+        w.set_pg(true);
+        // w.set_psize(2); // 32 bits at once
```
Comment (Luctins) [inline embassy-stm32/src/flash/h5.rs:65]: check if size can/has to be adjusted

File: embassy-stm32/src/flash/h5.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    loop {
+        let sr = pac::FLASH.nssr().read();
+        if !sr.bsy() && !sr.dbne() {
+            break;
+        }
+    }
+    clear_all_err();
+
+    pac::FLASH.nscr().modify(|r| {
+        // TODO: later check bank swap
```
Comment (Luctins) [inline embassy-stm32/src/flash/h5.rs:114]: is it relevant to implement the flash bank swapping logic?

File: embassy-stm32/src/flash/h5.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        r.set_bksel(match sector.bank {
```
Comment (Luctins) [inline embassy-stm32/src/flash/h5.rs:115]: is there a better way to map this? like some implicit conversion. if the bank swap logic is used this has to be inverted too.

File: embassy-stm32/src/flash/h5.rs [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    pac::FLASH.nscr().write(|w| w.set_pg(false));
+
+    unwrap!(res)
+}
+
+pub(crate) unsafe fn blocking_erase_sector(sector: &FlashSector) -> Result<(), Error> {
+    // pac::FLASH.wrp2r_cur().read().wrpsg()
+    // TODO: write protection check
+    if pac::FLASH.nscr().read().lock() == true {
+        error!("flash locked");
```
Comment (Luctins) [inline embassy-stm32/src/flash/h5.rs:102]: return error here as we cannot write when the flash is locked.

**Review by Luctins — COMMENTED** [review]
added some of the outstanding issues with the current implementation

**Comment by Luctins** [comment]
Can anyone review this?

**Comment by Luctins** [comment]
Of course.

**Comment by Luctins** [comment]
@Dirbaio I'm done fixing CI, could you review?

**Review by Dirbaio — APPROVED** [review]
thanks!

**Pull request merged**